packages/vm/src/state/baseStateManager.ts

```diff
@@ -0,0 +1,502 @@
+ * and we cannot guarantee a stable interface yet.
+ */
+export abstract class BaseStateManager {
+  _common: Common
+  _debug: Debugger
+  _cache!: Cache
+
+  _touched: Set<AddressHex>
+  _touchedStack: Set<AddressHex>[]
+  _originalStorageCache: Map<AddressHex, Map<AddressHex, Buffer>>
+
+  // EIP-2929 address/storage trackers.
+  // This maps both the accessed accounts and the accessed storage slots.
+  // It is a Map(Address => StorageSlots)
+  // It is possible that the storage slots set is empty. This means that the address is warm.
+  // It is not possible to have an accessed storage slot on a cold address (which is why this structure works)
+  // Each call level tracks their access themselves.
+  // In case of a commit, copy everything if the value does not exist, to the level above
+  // In case of a revert, discard any warm slots.
+  _accessedStorage: Map<string, Set<string>>[]
+
+  // Backup structure for address/storage tracker frames on reverts
+  // to also include on access list generation
+  _accessedStorageReverted: Map<string, Set<string>>[]
+
+  _checkpointCount: number
+
+  /**
+   * StateManager is run in DEBUG mode (default: false)
+   * Taken from DEBUG environment variable
+   *
+   * Safeguards on debug() calls are added for
+   * performance reasons to avoid string literal evaluation
+   * @hidden
+   */
+  protected readonly DEBUG: boolean = false
+
+  /**
+   * Needs to be called from the subclass constructor
+   */
+  constructor(opts: DefaultStateManagerOpts) {
+    let common = opts.common
+    if (!common) {
+      common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Petersburg })
+    }
+    this._common = common
+
+    this._touched = new Set()
+    this._touchedStack = []
+    this._originalStorageCache = new Map()
+
+    this._accessedStorage = [new Map()]
+    this._accessedStorageReverted = [new Map()]
+
+    this._checkpointCount = 0
+
+    // Safeguard if "process" is not available (browser)
+    if (process !== undefined && process.env.DEBUG) {
+      this.DEBUG = true
+    }
+    this._debug = createDebugLogger('vm:state')
+  }
+
+  /**
+   * Gets the account associated with `address`. Returns an empty account if the account does not exist.
+   * @param address - Address of the `account` to get
+   */
+  async getAccount(address: Address): Promise<Account> {
+    const account = await this._cache.getOrLoad(address)
+    return account
+  }
+
+  /**
+   * Saves an account into state under the provided `address`.
+   * @param address - Address under which to store `account`
+   * @param account - The account to store
+   */
+  async putAccount(address: Address, account: Account): Promise<void> {
+    if (this.DEBUG) {
+      this._debug(
+        `Save account address=${address} nonce=${account.nonce} balance=${
+          account.balance
+        } contract=${account.isContract() ? 'yes' : 'no'} empty=${account.isEmpty() ? 'yes' : 'no'}`
```

Comment:
This is nitpicky but why are we not just using `true/false` here as returned by the `isContract` and `isEmpty` methods instead of "yes/no"?  This is purely stylistic but true/false just feels like a more helpful reporting of the state of this account address than saying "yes/no".

Reply:
I found it a little bit more readable to have `yes/no` here (but brain needs to do a tiny bit less of internal mapping in this occasion 😛), but I have absolutely "no" opinion on things like that, happy with the one way or the other.

Will leave this now to not dismiss the review.